README.md
# node-unrar-js

[![Build Status](https://github.com/YuJianrong/node-unrar.js/workflows/Build/badge.svg)](https://github.com/YuJianrong/node-unrar.js/actions?query=workflow%3ABuild)
[![npm version](https://badge.fury.io/js/node-unrar-js.svg)](https://badge.fury.io/js/node-unrar-js)
[![MIT License](https://badges.frapsoft.com/os/mit/mit.svg?v=103)](https://opensource.org/licenses/mit-license.php)
[![TypeScript](https://badges.frapsoft.com/typescript/code/typescript.svg?v=101)](https://www.typescriptlang.org/)

`node-unrar.js` is a npm module to extract rar archive in pure JavaScript. It's combined by a Javascript adoption layer and JavaScript unRar library compiled by [Emscripten](http://emscripten.org/) from the [C++ unrar library](http://www.rarlab.com/rar/unrarsrc-6.1.7.tar.gz) which hosted on <http://www.rarlab.com/rar_add.htm> .

## Installation

You can install the module via `npm`:

```bash
npm install node-unrar-js
```

## Features

- Fully support for RAR archive, because it comes from the official source code.
- Unicode support, for both comment and file path/name.
- API for Extraction to both memory and file system.
- Both Commonjs module (for nodejs) and ES2015 module (for webpack) are supported.

## Unsupported Features

- Volume archives are not supported.
- Synchronize File System IO functions are used in File Extraction.

## API to create the extractor

- `async function createExtractorFromData(options: ExtractorFromDataOptions): Promise<Extractor<Uint8Array>>` - Create the in Memory Extractor

  Options `ExtractorFromDataOptions`:

  - `data: ArrayBuffer` : ArrayBuffer object of the RAR archive file
  - `password?: string` : _Optional_ password
  - `wasmBinary? ArrayBuffer;` : _Optional_ Use in browser/webpack, the wasm binary must be loaded in the code and send to this function to load the wasm code

- `async function createExtractorFromFile(options: ExtractorFromFileOptions): Promise<Extractor>` - Get the File Extractor

  Options `ExtractorFromFileOptions`:

  - `filepath: string` : File path of the RAR archive file
  - `targetPath?: string` : _Optional_ target folder
  - `password?: string` : _Optional_ password
  - `filenameTransform?: (filename: string) => string`: _Optional_ transform the file name before it's created on file system
  - `wasmBinary? ArrayBuffer;` : _Optional_ Use in nodejs/webpack, the wasm binary must be loaded in the code and send to this function to load the wasm code in webpack based nodejs project (please read [Used in Webpack-bundled NodeJS Project](#use-in-webpack-bundled-nodejs-project) for more details).

  _Node_: This function is not available in EM2015 Module since the EM2015 Module is used for webpack in Browser.

## API of the extractor

- `getFileList(): ArcList` : Get the file header and file list of the archive.

  Members in `ArcList`:

  - `arcHeader: ArcHeader` : The header of the archive
  - `fileHeaders: Generator<FileHeader>` : The iterator of the `FileHeader` objects

```js
{
  arcHeader: {
    comment: "",
    flags: {
      authInfo: false,
      headerEncrypted: false,
      lock: false,
      recoveryRecord: false,
      solid: false,
      volume: false,
    },
  }, fileHeaders: (Iterator)
    {
      crc: 0,
      flags: {
        directory: false,
        encrypted: false,
        solid: false,
      },
      method: "Storing",
      name: "FileName",
      packSize: 0,
      time: "2017-04-03T10:41:42.000",
      unpSize: 0,
      unpVer: "2.9",
      comment: "",
    },
}
```

- `extract(options: ExtractOptions)`: Extract the files.
  - Options `ExtractOptions`:
    - `files?: string[] | ((fileHeader: FileHeader) => boolean)` _Optional_ Extract all the files if `files` is empty
      - `string[]`: Extract the specific files only
      - `(fileHeader: FileHeader) => boolean`: Extract only the filtered file (_eg. extract only the files without password_).
    - `password?: string`: _Optional_ password for the extracted files only (Different password can be applied to any single file in RAR archive).
  - The return values are different for `createExtractorFromData` and `createExtractorFromFile`:
    - `ArcFiles<Uint8Array>` for `createExtractorFromData`
    - `ArcFiles` for `createExtractorFromFile`
  - Members in `ArcFiles`:
    - `arcHeader: ArcHeader` : The header of the archive
    - `files: Generator<ArcFile>` : The iterator of the `ArcFile` objects
    - Members in `ArcFile`:
      - `fileHeader: FileHeader` : The header of the extracted file
      - `extraction?: Uint8Array` : The extracted content of the file (`createExtractorFromData` only). If the `ArcFile` is a folder (`ArcFile.fileHeader.flags.directory` is `true`), the `extraction` will be undefined, otherwise it will be the content of the file (in `Uint8Array).

```js
{
  arcHeader: {...} // Same as the arcHeader above
  files: (Iterator)
    {
      fileHeader: {...} // Same as the fileHeader above
      extraction?: Uint8Array // createExtractorFromData only
  ]
}
```

**Note: The returned iterators from the two apis are lazy, it means the file header/content will not be parsed/extracted if any file is not iterated yet.**

## Exception

The customized Error Object `UnrarError` will be thrown for any exception in extracting. The definition of the Object is:

```Typescript
class UnrarError extends Error {
    reason: FailReason;
    file?: string | undefined; // Will be filled for any exception in extraction of a file
}
```

The following code is used in the `FailReason`:

| FailReason            | Message                                                    |
| --------------------- | ---------------------------------------------------------- |
| ERAR_NO_MEMORY        | Not enough memory                                          |
| ERAR_BAD_DATA         | Archive header or data are damaged                         |
| ERAR_BAD_ARCHIVE      | File is not RAR archive                                    |
| ERAR_UNKNOWN_FORMAT   | Unknown archive format                                     |
| ERAR_EOPEN            | File open error                                            |
| ERAR_ECREATE          | File create error                                          |
| ERAR_ECLOSE           | File close error                                           |
| ERAR_EREAD            | File read error                                            |
| ERAR_EWRITE           | File write error                                           |
| ERAR_SMALL_BUF        | Buffer for archive comment is too small, comment truncated |
| ERAR_UNKNOWN          | Unknown error                                              |
| ERAR_MISSING_PASSWORD | Password for encrypted file or header is not specified     |
| ERAR_EREFERENCE       | Cannot open file source for reference record               |
| ERAR_BAD_PASSWORD     | Wrong password is specified                                |

## Memory Leak

Note: although the return value `fileHeaders` and `files` are iterators, they must be traversed to the end! Otherwise the C++ object for archive extraction will not be destructed and cause memory leak.

## Example

```js
const fs = require("fs");
const unrar = require("node-unrar-js");

async function main() {
  // Read the archive file into a typedArray
  const buf = Uint8Array.from(fs.readFileSync("a.rar")).buffer;
  const extractor = await unrar.createExtractorFromData({ data: buf });

  const list = extractor.getFileList();
  const listArcHeader = list.arcHeader; // archive header
  const fileHeaders = [...list.fileHeaders]; // load the file headers

  const extracted = extractor.extract({ files: ["1.txt"] });
  // extracted.arcHeader  : archive header
  const files = [...extracted.files]; //load the files
  files[0].fileHeader; // file header
  files[0].extraction; // Uint8Array content, createExtractorFromData only
}
main();
```

## Demo in Webpack

This package can also be used in browser by Webpack, please visit the [demo project](https://github.com/YuJianrong/node-unrar.js/tree/master/demo/web) to see how to use it in webpack.

## Use in Webpack-bundled NodeJS Project

In most cases the exported ESModule is used in browser by Webpack, but in case if the NodeJs project (or an Electron project) is bundled by Webpack, the `wasmBinary` data must be loaded manually just like the browser by Webpack, it can be loaded like this:

```Typescript
import fs from 'fs';
import { createExtractorFromFile } from 'node-unrar-js/esm';

const wasmBinary = fs.readFileSync(require.resolve('node-unrar-js/esm/js/unrar.wasm'));

const extractor = await createExtractorFromFile({ wasmBinary, filepath: './WithComment.rar' });

// const list = extractor.getFileList();
```

Note: the package must be loaded from `'node-unrar-js/esm'` instead of `'node-unrar-js'` to enable the function `createExtractorFromFile` in ES Module.

## Use asm.js instead of WASM

In any case the WASM (WebAssembly) is not working for you, you may try to switch to [asm.js(http://asmjs.org/)). To use `asm.js`, please do:

```Typescript
// For CommonJS
const unrar = require("node-unrar-js/dist-asm");
// Or, For ESModule
import { createExtractorFromData } from 'node-unrar-js/esm-asm';

```

## TypeScript

This module is written in [TypeScript](https://www.typescriptlang.org/), you can import it directly in TypeScript and get the benefit of static type checking and auto-complete of IDE.

## Development and Contribution

If you want to compile the module by yourself, please follow the steps below:

- Install the [docker](https://www.docker.com/)

- Download the c++ source of unRar library by:

`npm run downloadUnrarSrc`

- Build for debug:

`npm run build:debug`

- Build for release

`npm run build:release`

- Run Test

`npm run test`

## License

This module is licensed under MIT.

### [Changelog](https://github.com/YuJianrong/node-unrar.js/blob/master/CHANGELOG.md)


CHANGELOG.md
# [2.1.0-beta.1](https://github.com/YuJianrong/node-unrar.js/compare/v2.0.2...v2.1.0-beta.1) (2023-11-29)


### Features

* enable asm.js ([976359b](https://github.com/YuJianrong/node-unrar.js/commit/976359b053e3df88ae56f36794a363a6c6e44b92))

## [2.0.2](https://github.com/YuJianrong/node-unrar.js/compare/v2.0.1...v2.0.2) (2023-11-28)


### Bug Fixes

* remove the postinstall script which should not be executed when installed on user side ([#329](https://github.com/YuJianrong/node-unrar.js/issues/329)) ([16b4e3c](https://github.com/YuJianrong/node-unrar.js/commit/16b4e3c0fca8602392d7f492830867b05e762f4a)), closes [#324](https://github.com/YuJianrong/node-unrar.js/issues/324)

## [2.0.1](https://github.com/YuJianrong/node-unrar.js/compare/v2.0.0...v2.0.1) (2023-11-22)


### Bug Fixes

* fix husky blocker on semantic-release ([88cd80c](https://github.com/YuJianrong/node-unrar.js/commit/88cd80cd60cb5dd054235e4a0376d2ad2b745213))
* fix the case when file size is greater than 2G ([#323](https://github.com/YuJianrong/node-unrar.js/issues/323)) ([f81cacd](https://github.com/YuJianrong/node-unrar.js/commit/f81cacde37e4725a453563e9faa58e7932c23dcb))

#### 2.0.0 (2022-06-15)

- Add support for NodeJs v18
- **(Breaking change)**: `ArcFile.extraction` is optional for `createExtractorFromData` now. If `ArcFile.fileHeader.flags.directory` is `true`, this field will be `undefined`.

#### 1.0.6 (2022-03-16)

- Add `createExtractorFromFile` support in Webpack

#### 1.0.5 (2022-02-28)

- Fix an issue on file extracting to current working directory ('./')

#### 1.0.4 (2022-02-15)

- Fix folder creation issue on Windows.

#### 1.0.3 (2021-05-10)

- Fix for Security Vulnerability on dependencies

#### 1.0.2 (2021-04-01)

- Fix for Security Vulnerability on dependencies

#### 1.0.1 (2021-02-18)

- Add new option `filenameTransform` for `ExtractorFromFileOptions`.

#### 1.0.0 (2020-12-29)

- Refactory by the latest EMScripten and UnRar DLL source. Since both of them changed a lot, there are many breaking changes in this update
- **(Breaking change)**: Use Javascript Exception instead of tuple `[state, result]` for error handling.
- **(Breaking change)**: The extractor object return by `createExtractorFromData` and `createExtractorFromFile` are now promise object. Since loading the wasm object in the latest EMScriten is asynchronized.
- **(Breaking change)**: The `fileHeaders` and `files` returned by the extractions api are changed from array to iterator, make the extraction lazy.

#### 0.8.0

- First release


src/js/extractor.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
export type SeekMethod = 'CUR' | 'SET' | 'END';

const ERROR_CODE = {
  0: 'ERAR_SUCCESS',
  10: 'ERAR_END_ARCHIVE',
  11: 'ERAR_NO_MEMORY',
  12: 'ERAR_BAD_DATA',
  13: 'ERAR_BAD_ARCHIVE',
  14: 'ERAR_UNKNOWN_FORMAT',
  15: 'ERAR_EOPEN',
  16: 'ERAR_ECREATE',
  17: 'ERAR_ECLOSE',
  18: 'ERAR_EREAD',
  19: 'ERAR_EWRITE',
  20: 'ERAR_SMALL_BUF',
  21: 'ERAR_UNKNOWN',
  22: 'ERAR_MISSING_PASSWORD',
  23: 'ERAR_EREFERENCE',
  24: 'ERAR_BAD_PASSWORD',
} as const;

export type FailReason = Exclude<
  typeof ERROR_CODE[keyof typeof ERROR_CODE],
  'ERAR_SUCCESS' | 'ERAR_END_ARCHIVE'
>;

const ERROR_MSG: { [k in FailReason]: string } = {
  ERAR_NO_MEMORY: 'Not enough memory',
  ERAR_BAD_DATA: 'Archive header or data are damaged',
  ERAR_BAD_ARCHIVE: 'File is not RAR archive',
  ERAR_UNKNOWN_FORMAT: 'Unknown archive format',
  ERAR_EOPEN: 'File open error',
  ERAR_ECREATE: 'File create error',
  ERAR_ECLOSE: 'File close error',
  ERAR_EREAD: 'File read error',
  ERAR_EWRITE: 'File write error',
  ERAR_SMALL_BUF: 'Buffer for archive comment is too small, comment truncated',
  ERAR_UNKNOWN: 'Unknown error',
  ERAR_MISSING_PASSWORD: 'Password for encrypted file or header is not specified',
  ERAR_EREFERENCE: 'Cannot open file source for reference record',
  ERAR_BAD_PASSWORD: 'Wrong password is specified',
};

export class UnrarError extends Error {
  constructor(public reason: FailReason, message: string, public file?: string) {
    super(message);
  }
}

export type CompressMethod =
  | 'Storing'
  | 'Fastest'
  | 'Fast'
  | 'Normal'
  | 'Good'
  | 'Best'
  | 'Unknown';

export interface FileHeader {
  name: string;
  flags: {
    encrypted: boolean;
    solid: boolean;
    directory: boolean;
  };
  packSize: number;
  unpSize: number;
  // hostOS: "MSDOS" | "OS/2" | "Win32" | "Unix",
  crc: number;
  time: string;
  unpVer: string;
  method: CompressMethod;
  comment: string;
  // fileAttr: number;
}

export interface ArcHeader {
  comment: string;
  flags: {
    volume: boolean;
    lock: boolean;
    solid: boolean;
    authInfo: boolean;
    recoveryRecord: boolean;
    headerEncrypted: boolean;
  };
  // files: FileHeader[];
}

export interface ArcList {
  arcHeader: ArcHeader;
  fileHeaders: Generator<FileHeader>;
}

export type ArcFile<withContent = never> = {
  fileHeader: FileHeader;
  extraction: withContent;
};

export interface ArcFiles<withContent = never> {
  arcHeader: ArcHeader;
  files: Generator<ArcFile<withContent>>;
}

export interface ExtractOptions {
  files?: string[] | ((fileHeader: FileHeader) => boolean);
  password?: string;
}

export abstract class Extractor<withContent = never> {
  protected abstract _filePath: string;
  private _password: string;
  private _archive: any;

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  constructor(protected unrar: any, password = '') {
    this._password = password;
    this._archive = null;
  }

  public getFileList(): ArcList {
    const arcHeader = this.openArc(true);

    function* getFileHeaders(this: Extractor<withContent>): Generator<FileHeader> {
      while (true) {
        const arcFile = this.processNextFile(() => true);
        if (arcFile === 'ERAR_END_ARCHIVE') {
          break;
        }
        yield arcFile.fileHeader;
      }
      this.closeArc();
    }

    return { arcHeader, fileHeaders: getFileHeaders.call(this) };
  }

  public extract({ files, password }: ExtractOptions = {}): ArcFiles<withContent> {
    const arcHeader = this.openArc(false, password);
    function* getFiles(this: Extractor<withContent>): Generator<ArcFile<withContent>> {
      let count = 0;
      while (true) {
        let shouldSkip: (fileHeader: FileHeader) => boolean = () => false;
        if (Array.isArray(files)) {
          if (count === files.length) {
            break;
          }
          shouldSkip = ({ name }: FileHeader) => !files.includes(name);
        } else if (files) {
          shouldSkip = (fileHeader) => !files(fileHeader);
        }
        const arcFile = this.processNextFile(shouldSkip);
        if (arcFile === 'ERAR_END_ARCHIVE') {
          break;
        }
        if (arcFile.extraction === 'skipped' || arcFile.fileHeader.flags.directory) {
          continue;
        }
        count++;
        yield {
          fileHeader: arcFile.fileHeader,
        } as ArcFile<withContent>;
      }
      this.closeArc();
    }
    return { arcHeader, files: getFiles.call(this) };
  }

  protected fileCreated(filename: string): void {
    return;
  }

  protected abstract open(filename: string): number;
  protected abstract create(filename: string): number;
  protected abstract read(fd: number, buf: number, size: number): number;
  protected abstract write(fd: number, buf: number, size: number): boolean;
  protected abstract tell(fd: number): number;
  protected abstract seek(fd: number, pos: number, method: SeekMethod): boolean;
  protected abstract closeFile(fd: number): void;

  protected close(fd: number): void {
    this.closeFile(fd);
  }

  private openArc(listOnly: boolean, password?: string): ArcHeader {
    this._archive = new this.unrar.RarArchive();
    const header = this._archive.open(
      this._filePath,
      password ? password : this._password,
      listOnly,
    );
    if (header.state.errCode !== 0) {
      throw this.getFailException(header.state.errCode, header.state.errType);
    }
    return {
      comment: header.comment,
      flags: {
        volume: (header.flags & 0x0001) !== 0,
        lock: (header.flags & 0x0004) !== 0,
        solid: (header.flags & 0x0008) !== 0,
        authInfo: (header.flags & 0x0020) !== 0,
        recoveryRecord: (header.flags & 0x0040) !== 0,
        headerEncrypted: (header.flags & 0x0080) !== 0,
      },
    };
  }

  private processNextFile(
    shouldSkip: (fileHeader: FileHeader) => boolean,
  ): ArcFile<'skipped' | 'extracted'> | 'ERAR_END_ARCHIVE' {
    function getDateString(dosTime: number): string {
      const bitLen = [5, 6, 5, 5, 4, 7];
      let parts: number[] = [];
      for (const len of bitLen) {
        parts.push(dosTime & ((1 << len) - 1));
        dosTime >>= len;
      }
      parts = parts.reverse();
      const pad = (num: number): string => (num < 10 ? '0' + num : '' + num);

      return (
        `${1980 + parts[0]}-${pad(parts[1])}-${pad(parts[2])}` +
        `T${pad(parts[3])}:${pad(parts[4])}:${pad(parts[5] * 2)}.000`
      );
    }

    function getMethod(method: number): CompressMethod {
      const methodMap: { [index: number]: CompressMethod } = {
        0x30: 'Storing',
        0x31: 'Fastest',
        0x32: 'Fast',
        0x33: 'Normal',
        0x34: 'Good',
        0x35: 'Best',
      };
      return methodMap[method] || 'Unknown';
    }

    const arcFileHeader = this._archive.getFileHeader();

    if (arcFileHeader.state.errCode === 10) {
      return 'ERAR_END_ARCHIVE';
    }

    if (arcFileHeader.state.errCode !== 0) {
      throw this.getFailException(arcFileHeader.state.errCode, arcFileHeader.state.errType);
    }

    const fileHeader: FileHeader = {
      name: arcFileHeader.name,
      flags: {
        encrypted: (arcFileHeader.flags & 0x04) !== 0,
        solid: (arcFileHeader.flags & 0x10) !== 0,
        directory: (arcFileHeader.flags & 0x20) !== 0,
      },
      packSize: arcFileHeader.packSize,
      unpSize: arcFileHeader.unpSize,
      // hostOS: arcFileHeader.hostOS
      crc: arcFileHeader.crc,
      time: getDateString(arcFileHeader.time),
      unpVer: `${Math.floor(arcFileHeader.unpVer / 10)}.${arcFileHeader.unpVer % 10}`,
      method: getMethod(arcFileHeader.method),
      comment: arcFileHeader.comment,
      // // fileAttr: arcFileHeader.fileAttr,
    };

    const skip = shouldSkip(fileHeader);
    const fileState = this._archive.readFile(skip);
    if (fileState.errCode !== 0) {
      throw this.getFailException(fileState.errCode, fileState.errType, fileHeader.name);
    }
    return {
      fileHeader,
      extraction: skip ? 'skipped' : 'extracted',
    };
  }

  private closeArc(): void {
    this._archive.delete();
    this._archive = null;
  }

  private getFailException(
    errCode: Exclude<keyof typeof ERROR_CODE, 0 | 10>,
    _errType: string,
    file?: string,
  ) {
    const reason = ERROR_CODE[errCode];
    this.closeArc();
    return new UnrarError(reason, ERROR_MSG[reason], file);
  }
}
